sshd/src/main/scala/ammonite/repl/sshd/ShellSession.scala

```diff
@@ -0,0 +1,63 @@
+package ammonite.repl.sshd
+
+import java.io.{InputStream, OutputStream}
+
+import ammonite.repl.sshd.util.Implicits
+import org.apache.sshd
+
+/**
+ * Implementation of ssh server's remote shell session, which will be serving remote user.
+ * @param remoteShell actual shell implementation, which will serving remote user's shell session.
+ */
+private[sshd] class ShellSession(remoteShell: SshServer.RemoteShellServer) extends sshd.server.Command {
+  var in: InputStream = _
+  var out: OutputStream = _
+  var exit: Option[sshd.server.ExitCallback] = None
+  var thread: Option[Thread] = None
+
+  def setInputStream(in: InputStream) = { this.in = in }
+  def setOutputStream(out: OutputStream) = { this.out = new SshOutputStream(out) }
+  def setErrorStream(err: OutputStream) = { }
+
+  // called by ssh server to instrument this session with a callback that it finished serving a user
+  def setExitCallback(exit: sshd.server.ExitCallback) { this.exit = Option(exit) }
+
+  // called then ssh server is ready to start this session. Starts the actual shell-serving task.
+  def start(env: sshd.server.Environment) {
+    import Implicits._
+    val thread = new Thread({
+      remoteShell(in, out)
+      this.thread = None
+      exit.foreach(_.onExit(0, "repl finished"))
+    })
+    this.thread = Some(thread)
+    thread.start()
+  }
+
+  // called then ssh server wants to destroy shell session. Whatever shell session serving a user
+  // was doing at this moment we are free to stop it.
+  def destroy() {
```

Comment:
No procedure syntax since it's deprecated https://issues.scala-lang.org/browse/SI-7605